utils/sp6_bitstream_analyzer.py

```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/env python3
+'''
+Spartan 6 bitstream analyzer tool.
+
+This script reads a Spartan6 bitstream and prints out some useful information.
+It can also create a frames file with the configuration data words.
+The bitstream is analyzed word by word and interpreted according to
+the UG380 Configuration User Guide.
+
+The tool can be used to derive the initialization, startup and finalization
+sequence as well as the configuration data. The latter is written to a frames
+file which can be used by the bitstream tools such as frames2bit to generate
+a valid bitstream.
+'''
+
+import argparse
+from io import StringIO
+
+conf_regs = {
+    0: "CRC",
+    1: "FAR_MAJ",
+    2: "FAR_MIN",
+    3: "FDRI",
+    4: "FDRO",
+    5: "CMD",
+    6: "CTL",
+    7: "MASK",
+    8: "STAT",
+    9: "LOUT",
+    10: "COR1",
+    11: "COR2",
+    12: "PWRDN_REG",
+    13: "FLR",
+    14: "IDCODE",
+    15: "CWDT",
+    16: "HC_OPT_REG",
+    18: "CSBO",
+    19: "GENERAL1",
+    20: "GENERAL2",
+    21: "GENERAL3",
+    22: "GENERAL4",
+    23: "GENERAL5",
+    24: "MODE_REG",
+    25: "PU_GWE",
+    26: "PU_GTS",
+    27: "MFWR",
+    28: "CCLK_FREQ",
+    29: "SEU_OPT",
+    30: "EXP_SIGN",
+    31: "RDBK_SIGN",
+    32: "BOOTSTS",
+    33: "EYE_MASK",
+    34: "CBC_REG"
+}
+
+cmd_reg_codes = {
+    0: "NULL",
+    1: "WCFG",
+    2: "MFW",
+    3: "LFRM",
+    4: "RCFG",
+    5: "START",
+    7: "RCRC",
+    8: "AGHIGH",
+    10: "GRESTORE",
+    11: "SHUTDOWN",
+    13: "DESYNC",
+    14: "IPROG"
+}
+
+opcodes = ("NOP", "READ", "WRITE", "UNKNOWN")
+
+
+def KnuthMorrisPratt(text, pattern):
+    '''Yields all starting positions of copies of the pattern in the text.
```

Comment:
This is not formatted correctly.

Reply:
Do you mean the docstring, dictionary or tuple? I do recall runinng format-py on this script...

Reply:
```python
def KnuthMorrisPratt(text, patter):
    ''''Yields all starting positions of copies of the pattern in the text.
    
    Calling conventions are similar to string.find, but its arguments can be
    lists or iterators, not just strings, it returns all matches, not just
    the first one, and it does not need the whole text in memory at once.
    Whenever it yields, it will have read the text exactly up to and including
    the match that caused the yield.
    '''
```

Reply:
Fixed.